plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,169 @@
-
-      extraFiles = mkIf cfg.nixvimInjections {
-        "queries/nix/injections.scm" = ''
-          ;; extends
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @lua)
-              (indented_string_expression (string_fragment) @lua)
-            ]
-            (#match? @_path "^extraConfigLua(Pre|Post)?$"))
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @vim)
-              (indented_string_expression (string_fragment) @vim)
-            ]
-            (#match? @_path "^extraConfigVim(Pre|Post)?$"))
-        '';
-      };
-
-      extraPlugins =
-        if cfg.nixGrammars then [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ] else [ cfg.package ];
-      extraPackages = with pkgs; [
-        tree-sitter
-        nodejs
-        cfg.gccPackage
-      ];
-
-      opts = mkIf cfg.folding {
-        foldmethod = "expr";
-        foldexpr = "nvim_treesitter#foldexpr()";
-      };
+    moduleConfig = helpers.defaultNullOpts.mkAttrsOf types.anything { } ''
+      This is the configuration for extra modules. It should not be used directly.
+    '';
+
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      description = "Install grammars with Nix";
     };
+
+    nixvimInjections = mkEnableOption "nixvim specific injections, like lua highlighting in extraConfigLua";
+  };
+
+  extraConfig = cfg: {
+    extraConfigLua =
+      (optionalString (cfg.parserInstallDir != null) ''
+        vim.opt.runtimepath:append("${cfg.parserInstallDir}")
+      '')
+      + ''
+        require('nvim-treesitter.configs').setup(${helpers.toLuaObject cfg.settings // cfg.moduleConfig})
+      ''
+      + (optionalString (cfg.languageRegister != { }) ''
+        __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
+
+        for parser_name, ft in pairs(__parserFiletypeMappings) do
+          require('vim.treesitter.language').register(parser_name, ft)
+        end
+      '');
+
+    extraFiles = mkIf cfg.nixvimInjections {
+      "queries/nix/injections.scm" = ''
+        ;; extends
+
+        (binding
+          attrpath: (attrpath (identifier) @_path)
+          expression: [
+            (string_expression (string_fragment) @lua)
+            (indented_string_expression (string_fragment) @lua)
+          ]
+          (#match? @_path "^extraConfigLua(Pre|Post)?$"))
+
+        (binding
+          attrpath: (attrpath (identifier) @_path)
+          expression: [
+            (string_expression (string_fragment) @vim)
+            (indented_string_expression (string_fragment) @vim)
+          ]
+          (#match? @_path "^extraConfigVim(Pre|Post)?$"))
+      '';
+    };
+
+    extraPlugins =
+      if cfg.nixGrammars then [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ] else [ cfg.package ];
```

Comment:
As a general best practice, prefer using mkMerge and mkIf in _config definitions_, unless there's a specific reason to use a "real" branch such as `if`, `optional`, etc.

This allows the module system to be a bit smarter; e.g. it'll check the definition is valid even when the condition is false

Reply:
Also, I think we'll need to add a makeNeovimPlugin argument that disables adding `cfg.package` to `extraPlugis`